guidance/rust_code/src/graph/mod.rs

```diff
@@ -1,402 +1,282 @@
-            let vertex = &graph.vertices[vertex_number];
-            if Some(&vertex.gps_point) == start_of_edge {
-                start_vertex_index = Some(vertex_number as usize);
+        for (vertex_index, vertex) in vertices.iter().enumerate() {
+            if vertex.point.gps == start_of_edge.gps {
+                start_vertex_index = Some(vertex_index);
             }
-            else if Some(&vertex.gps_point) ==  end_of_edge {
-                end_vertex_index = Some(vertex_number as usize);
+            else if vertex.point.gps ==  end_of_edge.gps {
+                end_vertex_index = Some(vertex_index);
             }
         }
 
         let start_vertex_index = start_vertex_index.unwrap();
         let end_vertex_index = end_vertex_index.unwrap();
 
-        graph.connection_matrix[start_vertex_index][end_vertex_index] = Some(edge_number);
-        graph.connection_matrix[end_vertex_index][start_vertex_index] = Some(edge_number);
+        connection_matrix[(start_vertex_index, end_vertex_index)] = Some(EdgeIndex(edege_index));
+        connection_matrix[(end_vertex_index, start_vertex_index)] = Some(EdgeIndex(edege_index));
     }
 
-    return graph
+    Graph {
+        edges,
+        vertices,
+        connection_matrix
+    }
 }
 
 
-pub fn initialize_from_kml_file(name: &str) -> Graph {
+pub fn initialize_from_gpx_file(name: &str) -> Graph {
+    // Open file and read contents to memory
     let file = File::open(name).unwrap();
-    // Open the file and store its contents to a buffer in RAM
-    let mut reader = BufReader::new(file);
-
-    let (number_of_edges, number_of_vertices) =
-        number_of_edges_and_vertices_from_buffer(&mut reader);
+    let reader = BufReader::new(file);
 
-    let (mut edges, mut vertices) = (
-        Vec::with_capacity(number_of_edges as usize),
-        Vec::with_capacity(number_of_vertices as usize),
-    );
+    let gpx_data = gpx::read(reader).unwrap();
 
-    add_gps_points_to_edges(&mut reader, &mut edges);
-    add_gps_points_to_vertices(&mut reader, &mut vertices);
+    let vertices = gpx_data.waypoints.into_iter()
+        .map(|vertex_data| {
+            // long, lat order is intentional. They are stored in this order in the file.
+            let (long, lat) = vertex_data.point().x_y();
 
+            let height = vertex_data.elevation.unwrap();
+            let name = vertex_data.name.unwrap();
+            
+            Vertex::new(name, GPSPointDeg{long, lat, height})
+        })
+        .collect::<Vec<Vertex>>();
+
+    let edges = gpx_data.tracks.into_iter() 
+        .map(|track| {
+            // Indexing at 0 since for every track we are guaranteed to only have move segment.
+            let points = track.segments[0].points.iter()
+                .map(|waypoint| {
+                    let height = waypoint.elevation.unwrap();
+                    let (long, lat) = waypoint.point().x_y();
+                    let gps = GPSPointDeg{long, lat, height};
```

Comment:
Same comment about GPS point wanting the order lat, long, height

Reply:
Does not matter the order in which I initialize, so long as the name of the parameters is correct. In this case, they are guaranteed to be correct since this is checked by the compiler.